### ISSUE TITLE
Crash pushing docs with non-downloaded attachments

### ISSUE THREAD

**Issue by EthanRTF** [couchbase/couchbase-lite-ios#1128]
CBL Version: dev-[e2f6c6f](https://github.com/couchbase/couchbase-lite-ios/commit/e2f6c6f7f384fa8b39020698fd4500078b8a42df)
Client OS: iOS 9.2
Server: 4.0.0-4051
Gateway: Release 1.2

I also encountered the error on the release/1.2.0 branch with the code from PR #1126 included.

During p2p replication, I receive the following error on multiple, but not all, iPads.

```
2016-02-17 17:25:41.807 posRetail[3212:1046883] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSArrayM insertObject:atIndex:]: object cannot be nil'
*** First throw call stack:
(0x183711900 0x182d7ff80 0x1835fb134 0x100a7c4bc 0x100a7bd70 0x100a7c578 0x100a5c66c 0x100a5c878 0x100a5a86c 0x100a4d11c 0x100a6dc0c 0x100a705cc 0x183e7a6ec 0x183e7a67c 0x183e7a7f0 0x183ce731c 0x183dd4e7c 0x102ed1bb0 0x102edb354 0x183cd3a88 0x1835f50ac 0x183cd396c 0x183cd382c 0x183cd3658 0x1836c8efc 0x1836c8910 0x1836c6690 0x1835f5680 0x183fed21c 0x100a4562c 0x1840d3c40 0x18337bb28 0x18337ba8c 0x183379028)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

After receiving this error, the app will crash. According to the debug navigator, it is occurring in [CBLMultiStreamWriter addInput:length:]:

![screen shot 2016-02-17 at 5 33 37 pm](https://cloud.githubusercontent.com/assets/16580518/13128613/aa7a7b4c-d59c-11e5-9b60-9e13db5b74e4.png)

Also, it may or may not be related but I have started receiving these warnings as well:

```
2016-02-17 17:18:06.411 posRetail[1049:331534] WARNING: Can't get binary data of attachment 'attachment' of {-Vhn8oqSvY9ouxOw8Eq-9Tw #2-d4d40f2ad57e2da9517a26b5c556b37a}
2016-02-17 17:18:06.412 posRetail[1049:331534] WARNING: CBLRestPusher[http://Pad2.local.:61537/juls2]: Couldn't expand attachments of {-Vhn8oqSvY9ouxOw8Eq-9Tw #2-d4d40f2ad57e2da9517a26b5c556b37a}
2016-02-17 17:18:06.414 posRetail[1049:331534] WARNING: Can't get binary data of attachment 'attachment' of {-ovGGvrPBzXdeZhdbIF_7Rl #2-f2632429d2cc52bffa356b02f982c0eb}
2016-02-17 17:18:06.415 posRetail[1049:331534] WARNING: CBLRestPusher[http://Pad2.local.:61537/juls2]: Couldn't expand attachments of {-ovGGvrPBzXdeZhdbIF_7Rl #2-f2632429d2cc52bffa356b02f982c0eb}
2016-02-17 17:18:06.417 posRetail[1049:331534] WARNING: Can't get binary data of attachment 'attachment' of {-5oC53HvR1bbHmugnLg2eXU #2-636f5dd971d07dad7beaebdf1aa5c866}
2016-02-17 17:18:06.418 posRetail[1049:331534] WARNING: CBLRestPusher[http://Pad2.local.:61537/juls2]: Couldn't expand attachments of {-5oC53HvR1bbHmugnLg2eXU #2-636f5dd971d07dad7beaebdf1aa5c866}
```

Other notable information:
- We are using ForestDB.
- This database has over 300,000 documents, 31,000 attachments with an average size of 1KB, with the complete database being over 400MB.
- This issue arose after fixing a memory issue #1017, but that could simply be chalked up to the fact that we weren't getting this far in the process.
- Sounds similar to issue #1124.

**Comment by snej** [comment]
The problem seems to be at CBLRestPusher.m:428:
```obj-c
            [bodyStream addStream: attachmentObj.contentStream length: attachmentObj->length];
```
Here the stream parameter (`attachmentObj.contentStream`) is unexpectedly nil, triggering the exception. It's certainly possible for `contentStream` to return nil, so I should add some code to guard against that.

**Comment by snej** [comment]
Are you setting `CBLReplication.downloadsAttachments` to `NO`? That might explain why attachment data is missing. Looks like there are areas of the code that don't expect a doc attachment to not be readable.

**Comment by twelvearrays** [comment]
Thanks for looking into this for us!  Our replicators download all attachments.  No where in our code do we set CBLReplication.downloadsAttachments to No.  Thanks!

**Comment by snej** [comment]
That's weird, then. Do you just have one client that ended up in this state, or is it reproducible?

**Comment by twelvearrays** [comment]
We had two clients end up in this state.  We were testing with a total of five clients with p2p replication and the other three seemed fine.  @EthanRTF please verify this.

We were planning on testing this more this week to see if we could reproduce it with other clients.

**Comment by EthanRTF** [comment]
That is correct @twelvearrays; It only affected two out of five. 

Hopefully with more testing this week we can get a solid way of reproducing the issue.

**Comment by snej** [comment]
FYI, the best way to report a backtrace is by entering `bt` in the debugger console and copying the output. This is lower-bandwidth, machine-readable, and will usually include source line numbers. Thanks.

**Comment by EthanRTF** [comment]
After testing for the last week, we were unable to reproduce the issue. Reviewing our commit history, I found that a team member was testing with ```CBLReplication.downloadsAttachments``` set to ```NO```. When we encountered the error, it was not set to ```NO```, but the iPads that were affected were the iPads that said member was using, which may have caused the issue to happen.

Until we are able to reproduce the issue or have any more findings, I am going to mark the issue as closed.
